### ISSUE TITLE
Investigate using Redis for distributed caching

### ISSUE THREAD

**Issue by billy1999** [sitewhere/sitewhere#756]
I encountered cache consistency issues , Consider global caching, such as redis

**Comment by derekadams** [comment]
We noticed the same thing while testing the 2.1 release and are in the process of adding a fix. The cache is currently wrapped around the client API made available for communicating with remote microservices, but this is a problem when using the REST API. For instance, call to `get` after `update` should always return the updated data. Our solution for the short term is to offer both a cached and uncached version of the APIs and the client microservice can choose which version to use. For the REST API, we will skip the cache for the primary entity you are interacting with, but use it for enrichment data. For instance, in the REST call for getting a device, the device data bypasses the cache, but the optional enrichment data for device type is loaded from cache.

We'll also look into adding Redis so that we can support distributed caching. In the monolithic version of the system (before we used K8s and Helm) the fact that we had to run a separate server for Redis was a non-starter, but with the Helm install it should be doable. We can look at running it as a sidecar for the services which require access to cached data. Thanks for reporting the issue.
